### PR TITLE
Streaming gateway

### PR DESCRIPTION
**Pull request by NassiHarel**
This change is [Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234)

### CODE REVIEW

**Review by NassiHarel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234)* status: 0 of 68 files reviewed, 1 unresolved discussion (waiting on @NassiHarel)

___
*[core/api-server/api/rest-api/internal/exec.js, line 15 at r6](https://reviewable.io/reviews/kube-hpc/hkube/1234#-MYsTjP12ziA6NyqOUYk:-MYsTjP12ziA6NyqOUYl:b-or5vu0) ([raw file](https://github.com/kube-hpc/hkube/blob/808471eb1c9cec0712a928336e055791a7edefd5/core/api-server/api/rest-api/internal/exec.js#L15)):*
> ```
> gatewayURLs
> ```
Don't always return gatewayURLs

File: core/api-server/lib/service/execution.js [hunk 1]
```diff
@@ -103,9 +104,24 @@ class ExecutionService {
             const statusObject = { timestamp: Date.now(), pipeline: pipeline.name, status: pipelineStatuses.PENDING, level: levels.INFO.name };
             await storageManager.hkubeIndex.put({ jobId }, tracer.startSpan.bind(tracer, { name: 'storage-put-index', parent: span.context() }));
             await stateManager.createJob({ jobId, userPipeline, pipeline: pipelineObject, status: statusObject });
+            const gatewayURLs = [];
+            await Promise.all(pipeline.nodes.map(async node => {
```
Comment (NassiHarel) [inline core/api-server/lib/service/execution.js:108]: use filter instead of if

Comment (NassiHarel) [inline core/api-server/lib/service/execution.js:108]: Move this to different function

File: core/api-server/lib/service/execution.js [hunk 2]
```diff
@@ -103,9 +104,24 @@ class ExecutionService {
...
+                    const gateURL = {};
+                    gateURL[node.nodeName] = `gateway/${name}`;
+                    gatewayURLs.push(gateURL);
+                }
+            }));
+
             await producer.createJob({ jobId, maxExceeded, parentSpan: span.context() });
             span.finish();
-            return jobId;
+            return { jobId, gatewayURLs };
```
Comment (NassiHarel) [inline core/api-server/lib/service/execution.js:124]: dont return gatewayURLs

File: core/api-server/lib/service/gateway.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    async _deleteGateway(options) {
+        const { name } = options;
+        const res = await stateManager.deleteGateway({ name });
+        const message = res.deleted === 0 ? 'deleted operation has failed' : 'deleted successfully';
+        return { message, name };
+    }
+
+    _hasCron(pipeline, gatewayName) {
```
Comment (NassiHarel) [inline core/api-server/lib/service/gateway.js:46]: remove this _hasCron

File: core/api-server/tests/exec-raw.js [hunk 4]
```diff
@@ -272,6 +272,29 @@ describe('Executions', () => {
             const response = await request(options);
             expect(response.body).to.have.property('jobId');
         });
+        it('should insert gateway', async () => {
```
Comment (NassiHarel) [inline core/api-server/tests/exec-raw.js:275]: you should add tests for 
- gateway without a name
- gateway not exists
- gateway already exists

File: core/api-server/lib/service/execution.js [hunk 5]
```diff
@@ -103,9 +104,24 @@ class ExecutionService {
...
+                    let name = node.spec?.gatewayName;
+                    const { description, mem } = node.spec && node.spec;
+                    if (!name) {
+                        name = `${jobId}-${node.nodeName}`;
+                    }
+                    await gatewayService.insertGateway({ name, description, mem, nodeName: node.nodeName, jobId });
+                    const gateURL = {};
+                    gateURL[node.nodeName] = `gateway/${name}`;
+                    gateways.push(gateURL);
+                }
```
Comment (NassiHarel) [inline core/api-server/lib/service/execution.js:119]: gateways.push({ [node.nodeName]: `gateway/${name}` });

File: core/api-server/lib/service/gateway.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+    async gatewaysList(options) {
+        const { sort, limit } = options;
+        return stateManager.getGateways({ sort, limit });
+    }
+
+    async deleteGateway(options) {
+        const { name: gatewayName } = options;
+        await validator.gateways.validateGatewayExists({ gatewayName });
+        await this._stopAllCrons(gatewayName);
```
Comment (NassiHarel) [inline core/api-server/lib/service/gateway.js:38]: remove not used code

**Review by yehiyam — COMMENTED** [review]
> Reviewed 1 of 37 files at r1, 1 of 81 files at r10.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234)* status: 2 of 122 files reviewed, 7 unresolved discussions (waiting on @golanha)

**Review by golanha — DISMISSED** [review]

**Review by NassiHarel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234)* status: 109 of 140 files reviewed, 12 unresolved discussions (waiting on @golanha, @NassiHarel, and @yehiyam)

___
*[core/algorithm-gateway/README.md, line 1 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KSwjN8klAZLLBbOhv:-M_VXmdR5e5IP1IV2oL1:b2nhswv) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/README.md#L1)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

I don't think this works anymore :-)
</blockquote></details>

Yep
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 1 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KT6HbCzsgUwA5cbp0:-M_VY8NQB-k524upRcWT:bmfwpm4) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L1)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

Why is this folder called algorithm_unique_folder? is it required by the wrapper?
</blockquote></details>

The `nodejs-wrapper` has a config like this:
```
config.algorithm = {
    path: process.env.ALGORITHM_PATH || 'algorithm_unique_folder',
    entryPoint: process.env.ALGORITHM_ENTRY_POINT || 'index.js'
};
```
I will have to override these defaults envs if I want to change the location
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 20 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KUPFw7XWGIQDfUM1X:-M_VYvDlExdnvWv4LP5O:b-9i7tpz) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L20)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

Can there be a race condition between `this._resolve = res` and `return this._resolve();`?
</blockquote></details>

I don't think so, because nodeJS is single-threaded
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 23 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KUGJv2aJzh0xQpLs1:-M_V_5ii2LZgmiQHd2qi:b-90ytlj) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L23)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

reject is not used
</blockquote></details>

Yes, for now
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 43 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KUeYf2ZewXACKI_3m:-M_Vlv84EYeT7s2xokug:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L43)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

Why throw -> catch -> throw?
</blockquote></details>

Done.
___
*[core/algorithm-gateway/lib/app.js, line 8 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KbkBMB_bsfKM2IOA2:-M_VbwJsEEQzEbP4VcuB:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/lib/app.js#L8)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

Is this just for unit tests? Maybe create a class so it can be a member or property?
</blockquote></details>

Done.
___
*[core/algorithm-operator/lib/reconcile/algorithm-gateway.js, line 13 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KdFQJ0jRnVPrKhve6:-M_VnKRv-yHhdszU75Yu:b-njpv6u) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-operator/lib/reconcile/algorithm-gateway.js#L13)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

The debug-worker has also `updated` test
`const updated = deployments.filter(d => _filterChangedDeployments(d, algorithms, versions, registry, clusterOptions, options));`
Maybe we need it here as well?
</blockquote></details>

The `updated` is related to `deployments`, but the `gateways` aren't `deployments`, they are `jobs` created inside the `task executor`
___
*[core/api-server/api/rest-api/internal/exec.js, line 18 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_Ki7T9CSvuvzzTqubb:-M_Vcg6t86RYJVH4ASvJ:b-mhf35d) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/api-server/api/rest-api/internal/exec.js#L18)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

> ```
> res.gateways
> ```
Is this used? Also is `res.jobId` used? I think we used it in the past for logging or something like that, but I can't find it in the code
</blockquote></details>

We are using it for logging, but the `res.gateways` is just Golan's copy-paste
___
*[core/api-server/lib/service/gateway.js, line 17 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KjYOi9fcycDjU5n_Y:-M_VdzbL0B0onfH54swI:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/api-server/lib/service/gateway.js#L17)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

> ```
> 'gateway'
> ```
use nodeKind.Gateway
</blockquote></details>

Done.
___
*[core/api-server/tests/exec-pipelines.js, line 21 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KmEen06W34YcPyAjX:-M_VkT6c3MrLoAg5ekCV:b-ta6yca) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/api-server/tests/exec-pipelines.js#L21)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

why double quotes? 
</blockquote></details>

There is no formatting for tests, fixing this....
___
*[core/worker/lib/states/stateManager.js, line 202 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_Knh4E0pf3NJ3aaCB2:-M_VkqzL4BQoRsxN40Pk:b2vghai) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/worker/lib/states/stateManager.js#L202)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

What is this if for? In what case does `done` gets called when in results state
</blockquote></details>

I don't remember the exact scenario, but this error of `invalid transition` happens all the time.
during stop or exit... you will see an error in dashboard `unhandledRejection: cannot read prop data of undefined`

**Review by yehiyam — CHANGES_REQUESTED** [review]
> Reviewed 31 of 31 files at r11, 1 of 1 files at r12, 2 of 2 files at r13.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234)* status: 140 of 141 files reviewed, 3 unresolved discussions (waiting on @NassiHarel and @yehiyam)

___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 1 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KT6HbCzsgUwA5cbp0:-M_Vqm9k78wCB6dcve5W:bqifx8o) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L1)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

The `nodejs-wrapper` has a config like this:
```
config.algorithm = {
    path: process.env.ALGORITHM_PATH || 'algorithm_unique_folder',
    entryPoint: process.env.ALGORITHM_ENTRY_POINT || 'index.js'
};
```
I will have to override these defaults envs if I want to change the location

</blockquote></details>

algorunner.Run accepts callback object that you can pass it the start method
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 20 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KUPFw7XWGIQDfUM1X:-M_Vr5gS8eg8IQ6Fg64A:b-qv71lk) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L20)):*
<details><summary><i>Previously, NassiHarel (Nassi Harel) wrote…</i></summary><blockquote>

I don't think so, because nodeJS is single-threaded
</blockquote></details>

` this._resolve = res;` is called in next tick

**Review by NassiHarel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234)* status: 137 of 141 files reviewed, 3 unresolved discussions (waiting on @golanha and @yehiyam)

___
*[core/algorithm-operator/lib/templates/algorithm-gateway.js, line 57 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KfQjVB9OvJShmu3SF:-M_W0K9JCvszWqZYbNLB:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-operator/lib/templates/algorithm-gateway.js#L57)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

shouldn't it be gatewayName?
</blockquote></details>

Done.
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 1 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KT6HbCzsgUwA5cbp0:-M_W0N5583QjhbAV-xsC:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L1)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

algorunner.Run accepts callback object that you can pass it the start method
</blockquote></details>

Done.
___
*[core/algorithm-gateway/algorithm_unique_folder/index.js, line 20 at r10](https://reviewable.io/reviews/kube-hpc/hkube/1234#-M_KUPFw7XWGIQDfUM1X:-M_VyT6C6R1JJ8rAljS9:b-z6su65) ([raw file](https://github.com/kube-hpc/hkube/blob/70e43ada559bf7bcb28477647fe886787f5d0670/core/algorithm-gateway/algorithm_unique_folder/index.js#L20)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

` this._resolve = res;` is called in next tick
</blockquote></details>

I think this code is synced (runs in one tick)
Can you verify this? because I tried to simulate this without success...


```
 setInterval(() => {
        events.emit('stop');
   }, 0);

 setTimeout(() => {
        events.emit('stop');
  }, 0);

 setImmediate(() => {
        events.emit('stop');
  });

 process.nextTick(() => {
        events.emit('stop');
  });

events.on('stop', () => {
         if (!this._resolve) {
            console.error('ooops');  // never happens
        }
        return this._resolve();
 });

return new Promise((res, rej) => {
        this._resolve = res;
        this._reject = rej;
    });
```

**Review by yehiyam — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r14, 3 of 3 files at r15.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/1234#-:-M_ZQL4u2srEmbXZCglN:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @NassiHarel)

**Pull request merged**